### PR TITLE
feat: Notification observation improvement [WPB-8776]

### PR DESCRIPTION
**Pull request by borichellow**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-8776" title="WPB-8776" target="_blank">StoryWPB-8776</a>  [Android] Enhance Notification Handling and Resource Efficiency
  </td></table>
  <br />
 

# What's new in this PR?

### Issues

Current implementation of observing the notifications is not perfect. GetNotification DB-query observes changes in `Message`, `User`, `Conversation`, `MessageAssetContent` and `MessageTextContent` DB-tables. So every-time something is changed there -> Flow emits a new value.

### Solutions

Instead of observing the notifications (using Flow in DAO), get notification once only when there are some messages that user should be notified about. For that: 

1.  Update `EphemeralEventsNotificationManager` (renamed into `NotificationEventsManager`) by adding funs:
    - `scheduleRegularNotificationChecking` for scheduling a re-checking of notifications (called when new message that user should be notified about was persisted)
    - `observeRegularNotificationsChecking` for observing the Flow that notifications were changed - need to re-check
2. Update `insertOrIgnoreMessage` in `MessageDAO` to return `InsertMessageResult` which is enum of: 
   - `INSERTED_NEED_TO_NOTIFY_USER`  - message was inserted and the conversation is not muted so user could be notified about it
   - `INSERTED_INTO_MUTED_CONVERSATION` - message was inserted and the conversation is muted so user shouldn't be notified about it
3. Update `PersistMessageUseCase` so after each succeed message persisting it checks if user should be notified about that message (if conversation is not mute, that is not "my message" and  the message is "notificationable") if so -> schedule the checking by `NotificationEventsManager`
4. In `GetNotificationsUseCase` instead of just observing `messageRepository.getNotificationMessage()` observe `NotificationEventsManager.observeRegularNotificationsChecking()` and re-check notifications only when it emits.

### CODE REVIEW

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/message/PersistMessageUseCase.kt [hunk 1]
```diff
@@ -34,16 +38,24 @@ interface PersistMessageUseCase {
...
-            updateConversationReadDate = message.isSelfTheSender(selfUserId),
+            updateConversationReadDate = isSelfSender,
             updateConversationModifiedDate = message.content.shouldUpdateConversationOrder()
-        )
+        ).onSuccess {
+            val isConversationMuted = it == InsertMessageResult.INSERTED_INTO_MUTED_CONVERSATION
+
+            if (!isConversationMuted && !isSelfSender && message.content.shouldNotifyUser()) {
+                notificationEventsManager.scheduleRegularNotificationChecking()
+            }
```
Comment (vitorhugods) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/message/PersistMessageUseCase.kt:57]: 🥇

**Review by vitorhugods — APPROVED** [review]
A great step forward.

-----

Regarding the architecture, I think some of our DAOs are a bit too clever.
Returning if the conversation was muted or not isn't really something that _insertMessage_ should do.

Maybe it shouldn't even update the `notificationDate` either, without being told by the parameters (_e.g._ `insertMessage(message, updateNotifications = true)`).

Whatever is inserting the message, in this case, `PersistMessageUseCase` (which could be renamed to `ConversationMessageHandler` or something like that), should be a centralised place where we could put all of this logic, and pass down to the DAOs and in-memory caches, instead of distributing the logic across the layers.

This is all something that I think we didn't care much about in the past and I don't think it needs to be addressed in this PR. 

Overall, nice to see this.

**Review by alexandreferris — APPROVED** [review]
Amazing work!

Just have a question that shouldn't block this PR but something we can think for the future on what other events we can show the notification for

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/message/PersistMessageUseCase.kt [hunk 2]
```diff
@@ -115,4 +127,58 @@ internal class PersistMessageUseCaseImpl(
...
+            MessageContent.ConversationVerifiedMLS,
+            MessageContent.ConversationDegradedProteus,
+            MessageContent.ConversationVerifiedProteus,
+            is MessageContent.Composite,
+            is MessageContent.ButtonAction,
+            is MessageContent.ButtonActionConfirmation,
+            is MessageContent.MemberChange.FederationRemoved,
+            is MessageContent.FederationStopped.ConnectionRemoved,
+            is MessageContent.FederationStopped.Removed,
+            is MessageContent.ConversationProtocolChanged,
```
[the remark this answers is not in the document]
Comment (borichellow) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/data/message/PersistMessageUseCase.kt:177]: I was comparing to DB-query, there we have 
```
content_type IN ('TEXT', 'RESTRICTED_ASSET', 'ASSET', 'KNOCK', 'MISSED_CALL', 'LOCATION')
```
So yeah, for no notifications about protocol or name changes :)

**Comment by borichellow** [comment]
Please do not merge it yet, i'd like @MohamadJaara  to check it first, as he's inspirer of this changes :)

**Review by MohamadJaara — APPROVED** [review]
LGTM

**Pull request merged**